tools/crm_rule.c

```diff
@@ -87,175 +82,6 @@ mode_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError **e
-    xmlXPathObjectPtr xpathObj = NULL;
-    char *xpath = NULL;
-    int rc = pcmk_rc_ok;
-    int max = 0;
-
-    /* Rules are under the constraints node in the XML, so first find that. */
-    cib_constraints = pcmk_find_cib_element(data_set->input,
-                                            XML_CIB_TAG_CONSTRAINTS);
-
-    /* Get all rules matching the given ID which are also simple enough for us to check.
-     * For the moment, these rules must only have a single date_expression child and:
-     * - Do not have a date_spec operation, or
-     * - Have a date_spec operation that contains years= but does not contain moon=.
-     *
-     * We do this in steps to provide better error messages.  First, check that there's
-     * any rule with the given ID.
-     */
-    xpath = crm_strdup_printf("//rule[@id='%s']", rule_id);
-    xpathObj = xpath_search(cib_constraints, xpath);
-    max = numXpathResults(xpathObj);
-
-    if (max == 0) {
-        rc = ENXIO;
-        g_set_error(&error, PCMK__RC_ERROR, rc, "No rule found with ID=%s", rule_id);
-        goto done;
-    } else if (max > 1) {
-        rc = ENXIO;
-        g_set_error(&error, PCMK__RC_ERROR, rc, "More than one rule with ID=%s found", rule_id);
-        goto done;
-    }
-
-    free(xpath);
-    freeXpathObject(xpathObj);
-
-    /* Next, make sure it has exactly one date_expression. */
-    xpath = crm_strdup_printf("//rule[@id='%s']//date_expression", rule_id);
-    xpathObj = xpath_search(cib_constraints, xpath);
-    max = numXpathResults(xpathObj);
-
-    if (max != 1) {
-        rc = EOPNOTSUPP;
-        g_set_error(&error, PCMK__RC_ERROR, rc,
-                    "Can't check rule %s because it does not have exactly one date_expression", rule_id);
-        goto done;
-    }
-
-    free(xpath);
-    freeXpathObject(xpathObj);
-
-    /* Then, check that it's something we actually support. */
-    xpath = crm_strdup_printf("//rule[@id='%s']//date_expression[@operation!='date_spec']", rule_id);
-    xpathObj = xpath_search(cib_constraints, xpath);
-    max = numXpathResults(xpathObj);
-
-    if (max == 0) {
-        free(xpath);
-        freeXpathObject(xpathObj);
-
-        xpath = crm_strdup_printf("//rule[@id='%s']//date_expression[@operation='date_spec' and date_spec/@years and not(date_spec/@moon)]",
-                                  rule_id);
-        xpathObj = xpath_search(cib_constraints, xpath);
-        max = numXpathResults(xpathObj);
-
-        if (max == 0) {
-            rc = ENXIO;
-            g_set_error(&error, PCMK__RC_ERROR, rc,
-                        "Rule either must not use date_spec, or use date_spec with years= but not moon=");
-            goto done;
-        }
-    }
-
-    match = getXpathResult(xpathObj, 0);
-
-    /* We should have ensured both of these pass with the xpath query above, but
-     * double checking can't hurt.
-     */
-    CRM_ASSERT(match != NULL);
-    CRM_ASSERT(find_expression_type(match) == time_expr);
```

Comment:
`time_expr` was not defined anywhere so I dropped it in when moving to libpacemaker

Reply:
? It's in include/crm/pengine/rules.h

Reply:
Ouch, yeah thank you. I saw this below, with `time_expr` in the doxygen block of the helper function when the argument was really called `expr`, and blanked out. It wouldn't have compiled if `time_expr` weren't defined.
```
 * \param[in]  time_expr    date_expression XML
...
 */
static int
eval_date_expression(xmlNode *expr, crm_time_t *now, crm_time_t *next_change)
```

Will fix.